### ISSUE TITLE
winrt::array_view cannot convert from gsl::span or other such representations

### ISSUE THREAD

**Issue by miniksa** [microsoft/xlang#620]
Trying to use something in an already safe structure like a gsl::span<T> or a std::basic_string_view<T> (until C++20's std::span<T>) should be easily convertable to winrt::array_view<T>.

I've found it is not.

winrt::array_view<T> seems to accept either a bare array, a std::array<T>, or a std::vector<T>. 

While I'd like it to naturally accept the gsl::span<T> or std::basic_string_view<T>, that might be difficult. But I had expected to at least be able to pass the .data() and .size() elements directly over to the other form.

However, that particular constructor is hidden from public view on winrt::array_view<T>. 

This represents a request to either make that constructor public so I can just hand over the data and size members to effectively yet-another-span type or to implement constructors accepting conversions from GSL and STD types that are spans as well.

**Comment by DefaultRyan** [comment]
`array_view` has a constructor that takes a pair of pointers. It's not as convenient as accepting `gsl::span` or `std::span` directly, but it's not much work to write up a converter.

Nevertheless, I agree, I've had to interop with `gsl::span` a couple times, and it would sure be nice to have `array_view` directly accept it. Unfortunately, the only conditional compilation hook I can see in GSL to light up cppwinrt behavior is their include guards, e.g. `GSL_SPAN_H`. Not sure if that's the kind of implementation-detail dependency cppwinrt should depend upon.

**Comment by DHowett-MSFT** [comment]
Can you do SFINAE given a `template <typename T>` by trying to use `t.data()` and `t.size()`? It seems like you could adapt to anything that has a data and size without having to specialize for GSL specifically.

**Comment by miniksa** [comment]
I was sort of hoping that we would be able to template the usage of anything with a .data() and .size() or alternatively make the constructor that's already there accept those.

I realize that the implementation details of other span libraries would be tough to depend on.

My workaround is to use .data() and .data() + .size() as the first and last, but that makes cppcorechecks supremely mad because now I'm doing unchecked pointer math. And on top of that, the winrt::array_view is about to just take the fruits of my pointer math and reverse it to stuff back into the hidden pointer/size constructor in the first place which was just a total waste of time.

**Comment by kennykerr** [comment]
Yep, the plan is to adopt `std::span` once it is available in the standard library and eventually remove `winrt::array_view` entirely.
